### ISSUE TITLE
jaxlib segfaults on macOS virtual environments but not on local machines

### ISSUE THREAD

**Issue by matthewfeickert** [actions/virtual-environments#3679]
### Description

This is a complimentary issue to https://github.com/scikit-hep/pyhf/issues/1501 and https://github.com/google/jax/issues/7128. We have observed that on macOS 10.15 jobs that the latest version of `jaxlib` (`jaxlib` `v0.1.68`) will segfault. However, this behavior is [NOT replicable on a local MacBook Air or MacBook Pro](https://github.com/scikit-hep/pyhf/issues/1501#issuecomment-870071961). This segfault is _only_ seen for `jaxlib` `v0.1.68` and not for `jaxlib` `v0.1.67`, which is strange, but given that it can not be reproduced on local machines seems to indicate that there is an environmental problem.

There is a public standalone minimal failing example here https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example

We are happy to help debug as much as possible.

cc @lukasheinrich @kratsg @jpivarski

### Virtual environments affected

- [ ] Ubuntu 16.04
- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [X] macOS 10.15
- [ ] macOS 11
- [ ] Windows Server 2016
- [ ] Windows Server 2019

### Image version and build link

* Version: `20210620.1`
* Public standalone example: https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example
* Specific example of a failed build: https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example/runs/2956373891

### Is it regression?

Unclear. 

### Expected behavior

In the minimal failing example https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example the behavior of the macOS builds should have been to succeed in the same manner. That being run

```yaml
name: CI

on:
  push:
  workflow_dispatch:

jobs:
  test:

    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, macos-latest]
        python-version: [3.7, 3.8, 3.9]

    steps:
    - uses: actions/checkout@v2

    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v2
      with:
        python-version: ${{ matrix.python-version }}

    - name: Install dependencies
      run: |
        python -m pip --no-cache-dir install --upgrade pip setuptools wheel
        python -m pip --no-cache-dir install -r requirements.txt

    - name: List installed Python packages
      run: |
        python -m pip list

    - name: printenv for debugging
      run: |
        printenv

    - name: JAX debug
      run: |
        python debug_32b.py
        python debug_64b.py
```

runs without the macOS jobs segfaulting.

### Actual behavior

The `jaxlib` minimal examples the macOS jobs segault and fail while the Ubuntu jobs pass

[![standalone_fail](https://user-images.githubusercontent.com/5142394/124029589-ebe91e80-d9ba-11eb-9e8a-6057073c1bea.png)](https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example/runs/2956373891)

c.f. https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example/runs/2956373891

### Repro steps

# Short answer

Trigger the CI in https://github.com/matthewfeickert/jaxlib-gha-minimal-failing-example

# Long answer

(This is the same, just relevant files copied)

With the following files:

```
# requirements.txt
jax~=0.2.16
jaxlib>=0.1.68
```

```python
# debug_32.py
import jax  # noqa: F401
import jax.numpy as jnp

print(jnp.asarray([-2, -1], dtype=jnp.float32))
print(jnp.asarray([-2, -1], dtype=jnp.float64))
```

```yaml
# .github/workflows/ci.yml
name: CI

on:
  push:
  workflow_dispatch:

jobs:
  test:

    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, macos-latest]
        python-version: [3.7, 3.8, 3.9]

    steps:
    - uses: actions/checkout@v2

    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v2
      with:
        python-version: ${{ matrix.python-version }}

    - name: Install dependencies
      run: |
        python -m pip --no-cache-dir install --upgrade pip setuptools wheel
        python -m pip --no-cache-dir install -r requirements.txt

    - name: List installed Python packages
      run: |
        python -m pip list

    - name: printenv for debugging
      run: |
        printenv

    - name: JAX debug
      run: |
        python debug_32b.py
```

start a GHA workflow job.

**Comment by dibir-magomedsaygitov** [comment]
Hello @matthewfeickert. Thank you for your report. We will take a look.

**Comment by miketimofeev** [comment]
@matthewfeickert  I've done some experiments and narrowed down the issue a bit. The `segmentation fault` error occurs only on VMs with a core count of less than 4 (we have 3-core VMs for GitHub actions at the moment) with any python version, even with the one installed from homebrew. VMs with 2 or 3 cores have `segmentation fault` error in 50-70% of cases, however, there was no single occurrence for 4-core VM.
I hope this helps as, unfortunately, we don't have much expertise in python debugging.

**Comment by matthewfeickert** [comment]
Thanks very much for this detailed study and information, @miketimofeev &mdash; truly appreciate it! I'm going to cross post this on https://github.com/google/jax/issues/7128 to make sure the JAX dev team gets this information as well and see if they can do something more actionable with it then I can.

**Comment by miketimofeev** [comment]
@matthewfeickert if we can help in any other way, feel free to ping us. 
PS: strangely that Ubuntu VMs also have 2 cores but don't have such an issue, so it seems to be macOS or underlying virtualization specific.

**Comment by matthewfeickert** [comment]
This has been resolved by the release of `jaxlib` `v0.1.69` (c.f. https://github.com/google/jax/issues/7128).

Thank you very much @miketimofeev for the time and effort that you put into debugging this as it definitely helped get this resolved faster!
